[SPARK-5108][BUILD] Jackson dependency management for Hadoop-2.6.0 support

zhzhan opened the pull request:
There is dependency compatibility issue. Currently hadoop-2.6.0 use  1.9.13 for jackson. Upgrade to the same version to make it consistent.

zhzhan committed:
test

zhzhan committed:
test

zhzhan committed:
test

zhzhan committed:
revert

zhzhan committed:
revert

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/zhzhan/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/zhzhan/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
Merge branch 'master' of https://github.com/apache/spark

zhzhan committed:
hadoop2.6 dependency management fix

SparkQA commented:
[Test build #25111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25111/consoleFull) for   PR 3914 at commit [`7048124`](https://github.com/apache/spark/commit/7048124dae63220daa45ce58ffe53a25eb0970f3).
 * This patch merges cleanly.

srowen commented on pom.xml:
```diff
@@ -149,7 +149,7 @@
     <jline.version>${scala.version}</jline.version>
     <jline.groupid>org.scala-lang</jline.groupid>
-    <jackson.version>1.8.8</jackson.version>
+    <jackson.version>1.9.13</jackson.version>
```
We have to be a little careful about this, since this will affect Jackson versions for all builds. I took a look, with `mvn -Pyarn -Dhadoop.version=2.2.0 -Dverbose dependency:tree | grep org.codehaus.jackson`, and it looks like even this older version of Hadoop brings in generally 1.9.x versions of Jackson. Confusingly, I see Hadoop bringing in 1.8.3, 1.9.11, and 1.9.13, I believe. Going back further, `mvn -Dverbose dependency:tree | grep org.codehaus.jackson` suggests that there was some use of Jackson 1.0.1 in ancient Hadoop 1.0.4.

No matter what, there are Jackson version inconsistencies. Jackson is fairly backwards compatible, so it appears that for example we've gotten away with managing the version up to at least 1.8.8. I'm inclined to say just go with this, indeed, to better match even moderately old Hadoop distributions.

Certainly worth testing a few combinations!

srowen commented on pom.xml:
```diff
@@ -830,7 +830,17 @@
+        <artifactId>jackson-jaxrs</artifactId>
+        <version>${jackson.version}</version>
+      </dependency>
+   </dependencies>
```
Tiny nit: this needs to indent one more space.

tgaloppo committed:
SPARK-5017 [MLlib] - Use SVD to compute determinant and inverse of covariance matrix

MultivariateGaussian was calling both pinv() and det() on the covariance matrix, effectively performing two matrix decompositions.  Both values are now computed using the singular value decompositon. Both the pseudo-inverse and the pseudo-determinant are used to guard against singular matrices.

Author: Travis Galoppo <tjg2107@columbia.edu>

Closes #3871 from tgaloppo/spark-5017 and squashes the following commits:

383b5b3 [Travis Galoppo] MultivariateGaussian - minor optimization in density calculation
a5b8bc5 [Travis Galoppo] Added additional points to tests in test suite. Fixed comment in MultivariateGaussian
629d9d0 [Travis Galoppo] Moved some test values from var to val.
dc3d0f7 [Travis Galoppo] Catch potential exception calculating pseudo-determinant. Style improvements.
d448137 [Travis Galoppo] Added test suite for MultivariateGaussian, including test for degenerate case.
1989be0 [Travis Galoppo] SPARK-5017 - Fixed to use SVD to compute determinant and inverse of covariance matrix.  Previous code called both pinv() and det(), effectively performing two matrix decompositions. Additionally, the pinv() implementation in Breeze is known to fail for singular matrices.
b4415ea [Travis Galoppo] Merge branch 'spark-5017' of https://github.com/tgaloppo/spark into spark-5017
6f11b6d [Travis Galoppo] SPARK-5017 - Use SVD to compute determinant and inverse of covariance matrix. Code was calling both det() and pinv(), effectively performing two matrix decompositions. Futhermore, Breeze pinv() currently fails for singular matrices.
fd9784c [Travis Galoppo] SPARK-5017 - Use SVD to compute determinant and inverse of covariance matrix

SparkQA commented:
[Test build #25111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25111/consoleFull) for   PR 3914 at commit [`7048124`](https://github.com/apache/spark/commit/7048124dae63220daa45ce58ffe53a25eb0970f3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25111/
Test PASSed.

viirya committed:
[SPARK-5050][Mllib] Add unit test for sqdist

Related to #3643. Follow the previous suggestion to add unit test for `sqdist` in `VectorsSuite`.

Author: Liang-Chi Hsieh <viirya@gmail.com>

Closes #3869 from viirya/sqdist_test and squashes the following commits:

fb743da [Liang-Chi Hsieh] Modified for comment and fix bug.
90a08f3 [Liang-Chi Hsieh] Modified for comment.
39a3ca6 [Liang-Chi Hsieh] Take care of special case.
b789f42 [Liang-Chi Hsieh] More proper unit test with random sparsity pattern.
c36be68 [Liang-Chi Hsieh] Add unit test for sqdist.

zhzhan committed:
indent fix

viirya committed:
[SPARK-5099][Mllib] Simplify logistic loss function

This is a minor pr where I think that we can simply take minus of `margin`, instead of subtracting  `margin`.

Mathematically, they are equal. But the modified equation is the common form of logistic loss function and so more readable. It also computes more accurate value as some quick tests show.

Author: Liang-Chi Hsieh <viirya@gmail.com>

Closes #3899 from viirya/logit_func and squashes the following commits:

91a3860 [Liang-Chi Hsieh] Modified for comment.
0aa51e4 [Liang-Chi Hsieh] Further simplified.
72a295e [Liang-Chi Hsieh] Revert LogLoss back and add more considerations in Logistic Loss.
a3f83ca [Liang-Chi Hsieh] Fix a bug.
2bc5712 [Liang-Chi Hsieh] Simplify loss function.

nchammas commented:
Also @zhzhan, could you squash your commits into one? Right now there appear to be a large number of merge commits that are not relevant to this PR. These will show up in the commit message body that gets generated if/when this PR gets merged.

SaintBacchus committed:
[YARN][SPARK-4929] Bug fix: fix the yarn-client code to support HA

Nowadays, yarn-client will exit directly when the HA change happens no matter how many times the am should retry.
The reason may be that the default final status only considerred the sys.exit, and the yarn-client HA cann't benefit from this.
So we should distinct the default final status between client and cluster, because the SUCCEEDED status may cause the HA failed in client mode and UNDEFINED may cause the error reporter in cluster when using sys.exit.

Author: huangzhaowei <carlmartinmax@gmail.com>

Closes #3771 from SaintBacchus/YarnHA and squashes the following commits:

c02bfcc [huangzhaowei] Improve the comment of the funciton 'getDefaultFinalStatus'
0e69924 [huangzhaowei] Bug fix: fix the yarn-client code to support HA

WangTaoTheTonic committed:
[SPARK-2165][YARN]add support for setting maxAppAttempts in the ApplicationSubmissionContext

...xt

https://issues.apache.org/jira/browse/SPARK-2165

I still have 2 questions:
* If this config is not set, we should use yarn's corresponding value or a default value(like 2) on spark side?
* Is the config name best? Or "spark.yarn.am.maxAttempts"?

Author: WangTaoTheTonic <barneystinson@aliyun.com>

Closes #3878 from WangTaoTheTonic/SPARK-2165 and squashes the following commits:

1416c83 [WangTaoTheTonic] use the name spark.yarn.maxAppAttempts
202ac85 [WangTaoTheTonic] rephrase some
afdfc99 [WangTaoTheTonic] more detailed description
91562c6 [WangTaoTheTonic] add support for setting maxAppAttempts in the ApplicationSubmissionContext

tsudukim committed:
[SPARK-2458] Make failed application log visible on History Server

Enabled HistoryServer to show incomplete applications.
We can see the log for incomplete applications by clicking the bottom link.

Author: Masayoshi TSUZUKI <tsudukim@oss.nttdata.co.jp>

Closes #3467 from tsudukim/feature/SPARK-2458-2 and squashes the following commits:

76205d2 [Masayoshi TSUZUKI] Fixed and added test code.
29a04a9 [Masayoshi TSUZUKI] Merge branch 'master' of github.com:tsudukim/spark into feature/SPARK-2458-2
f9ef854 [Masayoshi TSUZUKI] Added space between "if" and "(". Fixed "Incomplete" as capitalized in the web UI. Modified double negative variable name.
9b465b0 [Masayoshi TSUZUKI] Modified typo and better implementation.
3ed8a41 [Masayoshi TSUZUKI] Modified too long lines.
08ea14d [Masayoshi TSUZUKI] [SPARK-2458] Make failed application log visible on History Server

dbtsai committed:
[SPARK-5128][MLLib] Add common used log1pExp API in MLUtils

When `x` is positive and large, computing `math.log(1 + math.exp(x))` will lead to arithmetic
overflow. This will happen when `x > 709.78` which is not a very large number.
It can be addressed by rewriting the formula into `x + math.log1p(math.exp(-x))` when `x > 0`.

Author: DB Tsai <dbtsai@alpinenow.com>

Closes #3915 from dbtsai/mathutil and squashes the following commits:

bec6a84 [DB Tsai] remove empty line
3239541 [DB Tsai] revert part of patch into another PR
23144f3 [DB Tsai] doc
49f3658 [DB Tsai] temp
6c29ed3 [DB Tsai] formating
f8447f9 [DB Tsai] address another overflow issue in gradientMultiplier in LOR gradient code
64eefd0 [DB Tsai] first commit

suyanNone committed:
[SPARK-5132][Core]Correct stage Attempt Id key in stageInfofromJson

SPARK-5132:
stageInfoToJson: Stage Attempt Id
stageInfoFromJson: Attempt Id

Author: hushan[胡珊] <hushan@xiaomi.com>

Closes #3932 from suyanNone/json-stage and squashes the following commits:

41419ab [hushan[胡珊]] Correct stage Attempt Id key in stageInfofromJson

zhzhan committed:
squash all commits

zhzhan committed:
Merge branch 'master' into spark-5108

SparkQA commented:
[Test build #25172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25172/consoleFull) for   PR 3914 at commit [`d38b550`](https://github.com/apache/spark/commit/d38b55038a5f4188b5c7fd447be5a37cb3501323).
 * This patch merges cleanly.

zhzhan commented:
@nchammas I have created another PR https://github.com/apache/spark/pull/3937, and will close this one.

zhzhan commented:
@nchammas I accidentally close PR #3937. Have opened another one https://github.com/apache/spark/pull/3938 Sorry for the confusion.

zhzhan commented:
Create another PR https://github.com/apache/spark/pull/3938 removing confusing commits. Will close this one.

SparkQA commented:
[Test build #25172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25172/consoleFull) for   PR 3914 at commit [`d38b550`](https://github.com/apache/spark/commit/d38b55038a5f4188b5c7fd447be5a37cb3501323).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25172/
Test FAILed.
